File: fe-pc/ui/src/main/java/com/example/ui/config/ApiMethodsPrivate.java
package com.example.ui.config;

import com.example.ui.service.StorageService;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;

import org.json.JSONObject;

public class ApiMethodsPrivate {
    private static String token = StorageService.getToken();

    // Phương thức POST với token
    public static JSONObject postRequest(String urlString, JSONObject body) {
        try {
            // Tạo URL và mở kết nối
            URL url = new URL(urlString);
            HttpURLConnection conn = (HttpURLConnection) url.openConnection();

            // Thiết lập phương thức và header
            conn.setRequestMethod("POST");
            conn.setRequestProperty("Content-Type", "application/json; charset=UTF-8");
            if (token != null && !token.isEmpty()) {
                conn.setRequestProperty("Authorization", "Bearer " + token);
            }
            conn.setDoOutput(true);

            // Gửi dữ liệu JSON
            try (OutputStream os = conn.getOutputStream()) {
                byte[] input = body.toString().getBytes(StandardCharsets.UTF_8);
                os.write(input);
            }

            // Đọc phản hồi
            int statusCode = conn.getResponseCode();
            BufferedReader br = new BufferedReader(
                    new InputStreamReader(
                            statusCode >= 200 && statusCode < 300
                                    ? conn.getInputStream()
                                    : conn.getErrorStream(),
                            StandardCharsets.UTF_8));

            StringBuilder responseBuilder = new StringBuilder();
            String responseLine;
            while ((responseLine = br.readLine()) != null) {
                responseBuilder.append(responseLine.trim());
            }

            JSONObject jsonResponse = new JSONObject(responseBuilder.toString());

            // Xử lý response
            if (statusCode == 200 || statusCode == 201) {
                return jsonResponse;
            } else {
                JSONObject error = new JSONObject();
                error.put("status", "error");
                error.put("message", jsonResponse.optString("message", "Unknown error."));
                return error;
            }

        } catch (Exception e) {
            // In lỗi (nếu cần)
            System.err.println("API Error: " + e.getMessage());

            JSONObject error = new JSONObject();
            error.put("status", "error");
            error.put("message", "Cannot connect to the server. Please try again later: " + e.getMessage());
            return error;
        }
    }

    // Phương thức GET với token
    public static JSONObject getRequest(String urlString) {
        try {
            // Tạo URL và mở kết nối
            URL url = new URL(urlString);
            HttpURLConnection conn = (HttpURLConnection) url.openConnection();

            // Thiết lập phương thức và header
            conn.setRequestMethod("GET");
            conn.setRequestProperty("Content-Type", "application/json");
            if (token != null && !token.isEmpty()) {
                conn.setRequestProperty("Authorization", "Bearer " + token);
            }

            // Đọc phản hồi
            int statusCode = conn.getResponseCode();
            BufferedReader br = new BufferedReader(
                    new InputStreamReader(
                            statusCode >= 200 && statusCode < 300
                                    ? conn.getInputStream()
                                    : conn.getErrorStream(),
                            "utf-8"));

            StringBuilder responseBuilder = new StringBuilder();
            String responseLine;
            while ((responseLine = br.readLine()) != null) {
                responseBuilder.append(responseLine.trim());
            }

            JSONObject jsonResponse = new JSONObject(responseBuilder.toString());

            // Xử lý response
            if (statusCode == 200) {
                return jsonResponse;
            } else {
                JSONObject error = new JSONObject();
                error.put("status", "error");
                error.put("message", jsonResponse.optString("message", "Unknown error."));
                return error;
            }

        } catch (Exception e) {
            // In lỗi (nếu cần)
            System.err.println("API Error: " + e.getMessage());

            JSONObject error = new JSONObject();
            error.put("status", "error");
            error.put("message", "Cannot connect to the server. Please try again later: " + e.getMessage());
            return error;
        }
    }

    // Phương thức PUT với token
    public static JSONObject putRequest(String urlString, JSONObject body) {
        try {
            // Tạo URL và mở kết nối
            URL url = new URL(urlString);
            HttpURLConnection conn = (HttpURLConnection) url.openConnection();

            // Thiết lập phương thức và header
            conn.setRequestMethod("PUT");
            conn.setRequestProperty("Content-Type", "application/json");
            if (token != null && !token.isEmpty()) {
                conn.setRequestProperty("Authorization", "Bearer " + token);
            }
            conn.setDoOutput(true);

            // Gửi dữ liệu
            try (OutputStream os = conn.getOutputStream()) {
                byte[] input = body.toString().getBytes("utf-8");
                os.write(input, 0, input.length);
            }

            // Đọc phản hồi
            int statusCode = conn.getResponseCode();
            BufferedReader br = new BufferedReader(
                    new InputStreamReader(
                            statusCode >= 200 && statusCode < 300
                                    ? conn.getInputStream()
                                    : conn.getErrorStream(),
                            "utf-8"));

            StringBuilder responseBuilder = new StringBuilder();
            String responseLine;
            while ((responseLine = br.readLine()) != null) {
                responseBuilder.append(responseLine.trim());
            }

            JSONObject jsonResponse = new JSONObject(responseBuilder.toString());

            // Xử lý response
            if (statusCode == 200 || statusCode == 204) {
                return jsonResponse;
            } else {
                JSONObject error = new JSONObject();
                error.put("status", "error");
                error.put("message", jsonResponse.optString("message", "Unknown error."));
                return error;
            }

        } catch (Exception e) {
            // In lỗi (nếu cần)
            System.err.println("API Error: " + e.getMessage());

            JSONObject error = new JSONObject();
            error.put("status", "error");
            error.put("message", "Cannot connect to the server. Please try again later: " + e.getMessage());
            return error;
        }
    }

    // Phương thức DELETE với token
    public static JSONObject deleteRequest(String urlString) {
        try {
            // Tạo URL và mở kết nối
            URL url = new URL(urlString);
            HttpURLConnection conn = (HttpURLConnection) url.openConnection();

            // Thiết lập phương thức và header
            conn.setRequestMethod("DELETE");
            conn.setRequestProperty("Content-Type", "application/json");
            if (token != null && !token.isEmpty()) {
                conn.setRequestProperty("Authorization", "Bearer " + token);
            }

            // Đọc phản hồi
            int statusCode = conn.getResponseCode();
            BufferedReader br = new BufferedReader(
                    new InputStreamReader(
                            statusCode >= 200 && statusCode < 300
                                    ? conn.getInputStream()
                                    : conn.getErrorStream(),
                            "utf-8"));

            StringBuilder responseBuilder = new StringBuilder();
            String responseLine;
            while ((responseLine = br.readLine()) != null) {
                responseBuilder.append(responseLine.trim());
            }

            JSONObject jsonResponse = new JSONObject(responseBuilder.toString());

            // Xử lý response
            if (statusCode == 200 || statusCode == 204) {
                return jsonResponse;
            } else {
                JSONObject error = new JSONObject();
                error.put("status", "error");
                error.put("message", jsonResponse.optString("message", "Unknown error."));
                return error;
            }

        } catch (Exception e) {
            // In lỗi (nếu cần)
            System.err.println("API Error: " + e.getMessage());

            JSONObject error = new JSONObject();
            error.put("status", "error");
            error.put("message", "Cannot connect to the server. Please try again later: " + e.getMessage());
            return error;
        }
    }
}

File: fe-mobile/lifetipsui/app/src/main/java/com/example/lifetipsui/helper/ApiMethodsPrivate.kt
package com.example.lifetipsui.helper

import com.example.lifetipsui.service.StorageService
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import java.nio.charset.StandardCharsets

class ApiMethodsPrivate {
    companion object {

        private fun makeRequest(
            urlString: String,
            method: String,
            body: JSONObject? = null
        ): JSONObject {
            return try {
                val url = URL(urlString)
                val conn = url.openConnection() as HttpURLConnection
                conn.requestMethod = method
                conn.setRequestProperty("Content-Type", "application/json")

                // Gắn token nếu có
                val token = StorageService.getToken()
                token?.takeIf { it.isNotEmpty() }?.let {
                    conn.setRequestProperty("Authorization", "Bearer $it")
                }

                // Gửi body nếu có (POST/PUT)
                if (method == "POST" || method == "PUT") {
                    conn.doOutput = true
                    conn.outputStream.use { os ->
                        val input = body.toString().toByteArray(StandardCharsets.UTF_8)
                        os.write(input, 0, input.size)
                    }
                }

                // Đọc kết quả
                val statusCode = conn.responseCode
                val reader = BufferedReader(
                    InputStreamReader(
                        if (statusCode in 200..299) conn.inputStream else conn.errorStream,
                        StandardCharsets.UTF_8
                    )
                )

                val responseBuilder = StringBuilder()
                reader.forEachLine { responseBuilder.append(it.trim()) }
                val response = responseBuilder.toString()

                val jsonResponse = if (response.isNotEmpty()) JSONObject(response) else JSONObject()

                if (statusCode in 200..299) {
                    jsonResponse
                } else {
                    JSONObject().apply {
                        put("status", "error")
                        put("message", jsonResponse.optString("message", "Unknown error."))
                    }
                }

            } catch (e: Exception) {
                System.err.println("API Error: ${e.message}")
                println("API Error: ${e.message} - $e")
                JSONObject().apply {
                    put("status", "error")
                    put("message", "Cannot connect to the server. Please try again later: $e")
                }
            }
        }

        // GET request với dữ liệu query params
        fun getRequest(url: String, data: Map<String, String>? = null): JSONObject {
            println("Calling API URL: $url")
            val fullUrl = if (!data.isNullOrEmpty()) {
                val queryParams = data.entries.joinToString("&") { (key, value) ->
                    "${URLEncoder.encode(key, "UTF-8")}=${URLEncoder.encode(value, "UTF-8")}"
                }
                "$url?$queryParams"
            } else {
                url
            }
            return makeRequest(fullUrl, "GET")
        }

        fun postRequest(url: String, body: JSONObject): JSONObject =
            makeRequest(url, "POST", body)

        fun putRequest(url: String, body: JSONObject): JSONObject =
            makeRequest(url, "PUT", body)

        fun deleteRequest(url: String): JSONObject =
            makeRequest(url, "DELETE")
    }
}
